### PR TITLE
Allow to override non-visual glyph data properties

### LINKED ISSUE #2367: Different sizes not used for selection & nonselection glyphs
See the example here:

http://nbviewer.ipython.org/github/birdsarah/bokeh-miscellany/blob/master/notebooks/Different%20size%20for%20selected%20%26%20not%20selected%20demo.ipynb

Simplified for clarity:

``` python
    nonselection = Circle(x='x', y='y', fill_color='gray', size=5)
    selection = Circle(x='x', y='y', fill_color='red', size=25)
    plot.add_glyph(source, nonselection, selection_glyph=selection, nonselection_glyph=nonselection)
```

The size that is used is whatever is passed as the second argument to add_glyph.

### PR DESCRIPTION
**Pull request by mattpap**
This PR has the following goals:

1. Clean up definitions of data, visual and mapped properties of glyph views (set by `set_data()`, `set_visuals()` and `map_data()`), by removing all definitions that can auto-generated, fixing any issues and bugs along the way.
2. Allow derived glyphs to override non-visual data properties from the base glyph and/or allow derived glyphs to be unrelated to the base glyph, under the assumption that any hit testing, indexing, etc. will still be done only on the base glyph.

Example of a derived hover glyph with 2x the radius:

[Screencast from 28.11.2023 21:38:10.webm](https://github.com/bokeh/bokeh/assets/27475/f3fdb036-0b81-493f-acec-380819b1d234)

and with selection glyph unrelated to the base glyph:

image

Example usage:
```python
import numpy as np

from bokeh.core.properties import field
from bokeh.models import Rect
from bokeh.plotting import figure, show

N = 1000
x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 1.5
radii_big = radii * 2
colors = np.array([(r, g, 150) for r, g in zip(50+2*x, 30+2*y)], dtype="uint8")

p = figure()
p.add_tools("hover", "box_select")

gr = p.circle(x, y, radius=radii,
    fill_color=colors, fill_alpha=0.8, line_color=None, hover_fill_alpha=0.5)

gr.hover_glyph.radius = field("radii_big")
gr.data_source.data["radii_big"] = radii_big

gr.selection_glyph = Rect(
    line_color=None,
    fill_color=field("fill_color"),
    width=field("radii_big"),
    height=field("radius"),
)

show(p)
```

fixes #2367

### CODE REVIEW

**Comment by bryevdv** [comment]
```
export type Data = p.GlyphDataOf<Props>
```
👏 👏 👏

File: bokehjs/src/lib/models/renderers/glyph_renderer.ts [hunk 1]
```diff
@@ -298,6 +304,11 @@ export class GlyphRendererView extends DataRendererView {
     const glsupport = this.has_webgl
 
     this.glyph.map_data()
+    this.decimated_glyph.map_data()
+    this.selection_glyph.map_data()
+    this.nonselection_glyph.map_data()
+    this.hover_glyph?.map_data()
+    this.muted_glyph.map_data()
```
Comment (bryevdv) [inline bokehjs/src/lib/models/renderers/glyph_renderer.ts:311]: Can you speak about the implications of the current changes here and/or eventual plans? Naively, I would expect this to explode data size and computational overhead. Avoiding that duplication was the topic of https://github.com/bokeh/bokeh/issues/10403 It's possible the current PR accomplishes that already, and I am just missing it at a quick look, but then any pointers would be appreciated.

Comment (mattpap) [reply bokehjs/src/lib/models/renderers/glyph_renderer.ts:311]: Only properties that differ between a derived glyph (e.g. hover glyph) and the base will be computed (`set_data()`) and mapped (`map_data()`). Thus, for all existing setups, the only additional cost is iteration over properties and definition of getters that are used to fetch data from the parent glyph. So far I didn't notice any performance implications, which can be verified by observing the run time of bokehjs' integration tests. One thing I to still have address is custom definitions in `_set_data()` and `_map_data()`, which is the primary reason for the PR being WIP.

Comment (mattpap) [reply bokehjs/src/lib/models/renderers/glyph_renderer.ts:311]: > Avoiding that duplication was the topic of https://github.com/bokeh/bokeh/issues/10403

I overlooked that issue and it seems that I covered quite a few of the points there though not all.

Comment (bryevdv) [reply bokehjs/src/lib/models/renderers/glyph_renderer.ts:311]: The issue was just meant to spur discussion. The main concern was avoiding any that duplication so if we have already achieved that then things are in good shape 👍

**Comment by bryevdv** [comment]
Also, just FYI, IIRC there are some explicit admonitions in the docs about non-visual props not working on sub-glyphs, so it would be go to go through and remove those as well, once this PR is closer to completion.

**Comment by mattpap** [comment]
Although tests pass, I expect there will be a long tail of small tweaks needed as a follow-up, to handle corner cases and unlock full potential of this change in all glyphs. I would like this to be merged soon though, given the extent of changes in this PR.

**Review by bryevdv — APPROVED** [review]
OK, I had a _very_ brief look, mostly just scanning for any copy-pasta in the bulk updates. I won't claim to fully understand the details of the the new glyph inheritance. If you have a chance to write a couple of paragraphs about the design, that would be appreciated. 

Agree it's best to get this in sooner rather than later in 3.4 cycle. I'd like to see the following things along with this PR:

* new issue for 3.4 to update docs and choose some examples to show this new capability off with
* a dev build as soon as feasible so we can do a quick scan over rendered docs
